contents/security/crossSiteRequestForgery/crossSiteRequestForgery.md

```diff
@@ -0,0 +1,96 @@
+<input type="number" name="receiver-account-no" value="987654"/>
+<input type="number" name="amount" value="100000"/>
+<input type="submit" value="View more pictures"/>
+</form>
+```
+
+So, when she clicks that button, the browser actually sends the following request to `http://www.example-bank.com`:
+
+```
+POST http://www.example-bank.com/transfer HTTP/1.1
+Cookie: auth=1abcd2ek3292fsa390sdf
+
+receiver-name=Eve&receiver-account-no=987654&amount=100000
+```
+
+Since Alice is still logged in to `www.example-bank.com`, the browser automatically attaches the cookie `auth=1abcd2ek3292fsa390sdf` to any requests made to `www.example-bank.com`. Hence, the bank's server considers this as an authenticated request from Alice and proceeds to transfer $100,000 from Alice's account to account number 987654, which belongs to the attacker!
+
+## Conditions for a successful CSRF attack
+
+From the example, we can see that a CSRF attack works by forging a valid request which **inherits the identity and privileges of the victim** to the vulnerable web server. Such inheritance of privileges is possible because the browser attaches any credentials associated with a website to all requests made to the site. Therefore, if the user is currently authenticated to the site, the site will have no way to distinguish between the forged request sent by the victim and a legitimate request sent by the victim. In short, there are 3 conditions necessary for the attack:
+
+1. The target website is vulnerable to a CSRF attack (i.e. it cannot distinguish between forged and legitimate requests).
+2. The victim is authenticated to the target website.
+3. The victim is tricked to visit an untrusted website by social engineering.
+
+*Note: [Social engineering](https://en.wikipedia.org/wiki/Social_engineering_%28security%29) refers to psychological manipulation of people into performing actions. In our example, Eve shares a link with a description that matches Alice's interest, which increases the chance that Alice will follow the link. Thus, this action is a form of social engineering.*
+
+## Defense against CSRF attacks
+
+Defense against CSRF can come from both the website and the user. As the user, one can reduce the likelihood of being attacked by logging out of sensitive services like banking immediately after carrying out transactions. This will negate condition number 2. Condition number 3 can be mitigated if the user is extremely careful and never visit websites they do not know. However, users generally do not follow this and just click on interesting links they encounter on social media. Hence, this threat of social engineering could never be fully eliminated.
+
+For the web server, it can eliminate condition number 1 by adding a secret token as a hidden field to all forms that it serves. This token should be random and cannot be guessed by the attacker. Back to our example, the bank website should use a form like the following:
```

Comment:
Awkward expression; perhaps `As for the web server` or `On the server side, condition 1 can be eliminated by adding...`